sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp

```diff
@@ -0,0 +1,181 @@
+                "when Y is passed."
+            )
+        return {"VI": VI}
+    return {}
+
+{{for name_suffix, INPUT_DTYPE in (('64', 'DTYPE'),('32', 'np.float32'))}}
+
+from ._datasets_pair cimport DatasetsPair{{name_suffix}}
+
+
+cdef class PairwiseDistances{{name_suffix}}:
+    """float{{name_suffix}} implementation of PairwiseDistances."""
+
+    @classmethod
+    def compute(
+        cls,
+        X,
+        Y,
+        str metric="euclidean",
+        chunk_size=None,
+        dict metric_kwargs=None,
+        str strategy=None,
+    ):
+        """Compute the pairwise-distances matrix.
+
+        This classmethod is responsible for introspecting the arguments
+        values to dispatch to the most appropriate implementation of
+        :class:`PairwiseDistances{{name_suffix}}`.
+
+        This allows decoupling the API entirely from the implementation details
+        whilst maintaining RAII: all temporarily allocated datastructures necessary
+        for the concrete implementation are therefore freed when this classmethod
+        returns.
+
+        No instance should directly be created outside of this class method.
+        """
+        # Precompute data-derived distance metric parameters
+        metric_kwargs = {} if metric_kwargs is None else metric_kwargs
+
+        params = _precompute_metric_params(
+            X,
+            Y,
+            metric=metric,
+            **metric_kwargs,
+        )
+        metric_kwargs.update(**params)
+
+        # Fall back on a generic implementation that handles most scipy
+        # metrics by computing the distances between 2 vectors at a time.
+        pdr = PairwiseDistances{{name_suffix}}(
+            datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+            strategy=strategy,
+        )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pdr.execute_in_parallel_on_Y:
+                pdr._parallel_on_Y()
+            else:
+                pdr._parallel_on_X()
+
+        return pdr._finalize_results()
+
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        strategy=None,
+        sort_results=False,
+    ):
+        self.datasets_pair = datasets_pair
+        self.n_samples_X = datasets_pair.n_samples_X()
+        self.n_samples_Y = datasets_pair.n_samples_Y()
+
+        self.effective_n_threads = _openmp_effective_n_threads()
+
+        if strategy is None:
+            strategy = get_config().get("pairwise_dist_parallel_strategy", 'auto')
+
+        if strategy not in ('parallel_on_X', 'parallel_on_Y', 'auto'):
+            raise RuntimeError(f"strategy must be 'parallel_on_X, 'parallel_on_Y', "
+                               f"or 'auto', but currently strategy='{self.strategy}'.")
+
+        if strategy == 'auto':
```

Comment:
I think the heuristic must be adapted in this case. Yet, I think this better might done in another PR as a follow-up. Hence I propose adding a TODO to mention this.
```suggestion
        # TODO: inspect if the current heuristic is relevant
        # for PairwiseDistances
        if strategy == 'auto':
```